### PR TITLE
Annotations on generic or array incorrectly considered fix

### PR DESCRIPTION
**Pull request by gtoison**
The NonNull annotations might be on the return type or on the parameters but it might also be on generics of the corresponding types. When they are on the return/parameter we should consider them, when they are on the generics we (currently) should ignore them.

For instance `List<@NonNull V>` means that the list contains non null elements, but the list itself might be null

See #2500

### CODE REVIEW

**Review by ThrawnCA — APPROVED** [review]

**Comment by trancexpress** [comment]
Looks good, the bug is fixed with this PR.

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java [hunk 1]
```diff
@@ -16,11 +17,69 @@ public class TestCheckerFrameworkTypeAnnotations extends AbstractIntegrationTest
...
                 .build();
 
-        assertThat(getBugCollection(), containsExactly(1, matcher));
+        // Per-method NP_NONNULL_RETURN_VIOLATION matchers
+        final BugInstanceMatcher matcherReturningNullOnNonNullMethod = new BugInstanceMatcherBuilder()
+                .bugType("NP_NONNULL_RETURN_VIOLATION")
+                .inMethod("returningNullOnNonNullMethod")
+                .build();
+
+        final BugInstanceMatcher matcherReturningNullWithNonNullOnGenerics = new BugInstanceMatcherBuilder()
```
Comment (JuditKnoll) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java:32]: IMO it's better if the positive tests are as specific as possible, and contain any additional info, which is added to the bug instance (e.g. source line location, variable).

**Review by JuditKnoll — CHANGES_REQUESTED** [review]

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java [hunk 2]
```diff
@@ -16,11 +17,69 @@ public class TestCheckerFrameworkTypeAnnotations extends AbstractIntegrationTest
...
-        final BugInstanceMatcher matcher = new BugInstanceMatcherBuilder()
+
+        // A matcher for all NP_NONNULL_RETURN_VIOLATION
+        final BugInstanceMatcher matcherMethodReturns = new BugInstanceMatcherBuilder()
                 .bugType("NP_NONNULL_RETURN_VIOLATION")
                 .build();
 
-        assertThat(getBugCollection(), containsExactly(1, matcher));
+        // Per-method NP_NONNULL_RETURN_VIOLATION matchers
+        final BugInstanceMatcher matcherReturningNullOnNonNullMethod = new BugInstanceMatcherBuilder()
```
[the remark this answers is not in the document]
Comment (gtoison) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java]: Agreed, I've made some changes
Let me know if that's better or if further changes are needed

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java (hunk 1 above)
Comment (gtoison) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java]: I've added the line number, let me know if that works for you

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java]: This change groups the matchers and the asserts making the code easier to read, but doesn't really help with the code repetition. I originally meant having a few functions, which has the bugtype, the method name and the asserted number of matches as parameters, and calling these functions in the test.

[the remark this answers is not in the document]
Comment (gtoison) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java]: I see, some kind of ... constructor ... then :)
I suppose that defeats the purpose of the builder pattern (if not a big fan of it anyway), I can add some functions to that test class to reduce the code duplication
So let me know if I should make that change

[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/TestCheckerFrameworkTypeAnnotations.java]: I think we are misunderstanding each other. I'm sorry if I didn't express myself properly. Here is an example in the current codebase of what I had in mind: https://github.com/spotbugs/spotbugs/blob/master/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/DontUseFloatsAsLoopCountersTest.java

**Review by JuditKnoll — APPROVED** [review]
LGTM! It seems like this issue solves several older issues as well.

**Pull request merged**